bus-mapping/src/evm/opcodes/ternop.rs

```diff
@@ -0,0 +1,129 @@
+use super::Opcode;
+use crate::circuit_input_builder::CircuitInputStateRef;
+use crate::eth_types::GethExecStep;
+use crate::{
+    operation::{StackOp, RW},
+    Error,
+};
+
+/// Placeholder structure used to implement [`Opcode`] trait over it corresponding to all the
+/// ternary operations: take three words and return one.
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct TernaryOpcode;
+
+impl Opcode for TernaryOpcode {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        steps: &[GethExecStep],
+    ) -> Result<(), Error> {
+        let step = &steps[0];
+        // 3 stack reads
+        for i in 0..3 {
+            state.push_op(StackOp::new(
+                RW::READ,
+                step.stack.nth_last_filled(i),
+                step.stack.nth_last(i)?,
+            ));
+        }
+
+        // Get operator result from next step and do stack write
+        let result_value = steps[1].stack.last()?;
+        state.push_op(StackOp::new(
+            RW::WRITE,
+            step.stack.nth_last_filled(2),
+            result_value,
```

Comment:
It seems more intuitive to me to use next step's address directly.

```suggestion
            steps[1].stack.last_filled(),
            steps[1].stack.last()?,
```

Reply:
Running out of gas I think that conditions again the fact of taking the value from one place or the other.